### Model selection

In [1]:
import pandas as pd
import pickle

In [2]:
## read the saved data frames
path = "C:/Users/nikos/Desktop/models_df/"

files = ["knn_df.pickle", "mlr_df.pickle", "rf_df.pickle", "svm_df.pickle"]

df = pd.DataFrame()
for i in files:
    new_path = path + i
    
    with open(new_path, 'rb') as dt:
        new_df = pickle.load(dt)
        
    df = df.append(new_df)
df = df.reset_index().drop('index', axis=1)  

In [3]:
## Print the Models table
df

Model  Train Set Accuracy  Test Set Accuracy
0                               KNN            0.855208           0.804979
1  Multionomial Logistic Regression            0.910417           0.850622
2                     Random Forest            0.887500           0.775934
3                               SVM            0.884375           0.846473

#### The model that seems to generate better results on the training and test set is the SVM model. So I will use the SVM algorithm

In [4]:
## Load the model and the data frame in order to examine some misclassified observations
# SVM models
path_model = r"C:\Users\nikos\Desktop\models_df\best_svm.pickle"
with open(path_model, 'rb') as dt:
    svm_model = pickle.load(dt)

In [6]:
## Load the data
path_X_train = r"C:\Users\nikos\Desktop\Classify reviews\model_data\X_train.pickle"
with open(path_X_train, 'rb') as data:
    red_X_train = pickle.load(data)

path_X_test = r"C:\Users\nikos\Desktop\Classify reviews\model_data\X_test.pickle"
with open(path_X_test, 'rb') as data:
    red_X_test = pickle.load(data)

path_y_train = r"C:\Users\nikos\Desktop\Classify reviews\model_data\y_train.pickle"
with open(path_y_train, 'rb') as data:
    red_y_train = pickle.load(data)

path_y_test = r"C:\Users\nikos\Desktop\Classify reviews\model_data\y_test.pickle"
with open(path_y_test, 'rb') as data:
    red_y_test = pickle.load(data)
    
path_df = r"C:\Users\nikos\Desktop\Classify reviews\model_data\df.pickle"
with open(path_df, 'rb') as data:
     df = pickle.load(data)

In [7]:
predictions = svm_model.predict(red_X_test)
predictions

array([3, 3, 5, 7, 4, 5, 1, 4, 7, 5, 3, 0, 5, 6, 0, 0, 0, 4, 0, 7, 7, 4,
       4, 5, 4, 4, 6, 7, 3, 7, 3, 7, 5, 7, 3, 1, 1, 1, 0, 0, 5, 6, 0, 7,
       4, 2, 6, 6, 6, 1, 0, 3, 2, 7, 0, 4, 2, 3, 3, 4, 6, 6, 2, 4, 5, 7,
       3, 0, 0, 6, 2, 3, 7, 2, 4, 2, 7, 7, 7, 1, 3, 3, 0, 4, 6, 6, 3, 4,
       4, 0, 0, 3, 0, 3, 3, 5, 2, 1, 6, 5, 3, 2, 7, 1, 3, 1, 7, 2, 2, 3,
       0, 0, 7, 6, 3, 3, 1, 4, 2, 0, 6, 7, 4, 1, 3, 1, 3, 4, 3, 1, 2, 0,
       4, 1, 0, 3, 5, 3, 7, 6, 1, 3, 7, 5, 4, 7, 5, 7, 1, 2, 4, 0, 4, 5,
       0, 3, 6, 0, 1, 3, 6, 3, 3, 2, 6, 6, 4, 3, 4, 6, 3, 1, 1, 2, 7, 1,
       3, 4, 5, 5, 3, 6, 2, 3, 3, 1, 3, 6, 3, 1, 2, 3, 0, 6, 1, 2, 6, 3,
       3, 1, 5, 0, 2, 5, 1, 4, 6, 4, 3, 7, 7, 6, 7, 4, 6, 3, 1, 1, 7, 3,
       7, 4, 4, 7, 7, 1, 3, 4, 1, 4, 4, 1, 7, 1, 7, 1, 5, 6, 1, 5, 1],
      dtype=int64)

In [8]:
# df_X_test  
path_X_test = r"C:\Users\nikos\Desktop\Classify reviews\Pickles\df_X_test.pickle"
with open(path_X_test, 'rb') as data:
    path_X_test = pickle.load(data)

In [9]:
df1 = path_X_test
df2 = red_y_test
df3 = pd.DataFrame(predictions)
result = pd.concat([df1, df2], axis=1)
result["predictions"] = predictions
result

Desc proc  Target  predictions
243   drug analyst report directly therapeutic area ...       5            3
1024  full timetemporarywe look hard work reliable p...       3            3
233   facebook mission give people power build commu...       5            5
70    king college hospital nh foundation trust one ...       4            7
363   job introductionshaw healthcare currently recr...       0            4
...                                                 ...     ...          ...
83    post office assistantpost office clerk provide...       5            5
240   new data analyst join vashi grow inhouse analy...       6            6
622   bleecker start food truck today shop around lo...       1            1
319   job descriptionwe role experienced kitchen por...       5            5
129   company descriptiontempdent recruitment traini...       1            1

[241 rows x 3 columns]

In [10]:
encode = {0:'Admin', 1:'Analyst', 2:'cleaner', 3:'Finance', 4:'hospitality',
       5:'recruitment', 6:'Warehouse', 7:'web developer'}

# Decode
result = result.replace({'Target':encode})
result = result.replace({'predictions':encode})
result.head()

Desc proc       Target  \
243   drug analyst report directly therapeutic area ...  recruitment   
1024  full timetemporarywe look hard work reliable p...      Finance   
233   facebook mission give people power build commu...  recruitment   
70    king college hospital nh foundation trust one ...  hospitality   
363   job introductionshaw healthcare currently recr...        Admin   

        predictions  
243         Finance  
1024        Finance  
233     recruitment  
70    web developer  
363     hospitality

In [11]:
## Inspect some misclassified observations
condition = (result['Target'] != result['predictions'])

df_misclassified = result[condition]

df_misclassified.head()

Desc proc       Target  \
243   drug analyst report directly therapeutic area ...  recruitment   
70    king college hospital nh foundation trust one ...  hospitality   
363   job introductionshaw healthcare currently recr...        Admin   
879   establish consultancy seek recruitment consult...        Admin   
1032  fully conversant detail air freight forward wa...      cleaner   

        predictions  
243         Finance  
70    web developer  
363     hospitality  
879     hospitality  
1032        Finance

In [12]:
## read some of them 
df_misclassified["Desc proc"][243]

'drug analyst report directly therapeutic area director associate director member citeline editorial team drug analyst broadly responsible support daily editorial operation ensure quality integrity pharmaprojects database provide unparalleled research support client support analytics project objective achieve solid understanding varied competitive drug r intelligence source collect key drug development data conduct additional research devise basic database search strategy client perform basic data responsibilitiescontent productionsupport therapeutic area team maintain accurate database citeline productsreview update add drug record base company pipeline medical sec filing information sourcesacquire basic understanding therapeutic area include disease scope coverageacquire basic understanding competence internal content management system master database functionality process relate record creation maintenanceacquire basic knowledge understanding relationship citeline product ensure con

In [13]:
## This description is about an Analyst role therefore, the label and the prediction are not corresponding to the meaning

In [14]:
df_misclassified["Desc proc"][363]

'job introductionshaw healthcare currently recruit domestic supervisor preferably experience work busy environment housekeeper similar role work environment support vulnerable people important experience work spending time older people understand need carry variety cleaning task throughout home whatever require maintain highest level standard whilst supervise domestic responsibilitieswhat arrange allocation work duty cleaning staff ensure adequate staffing cleaning schedule schedule periodic cleaning conjunction service clean frequency ie daily weekly issue use cleaning safe use coshh regulate product ensure date coshh assessment form readily supervise develop domestic staffmonitor cleaning health safety standard take remedial action control record monitoring procedure place advise service manager facility manager regional manage carry effective communication care staff resident visitor respond need residentsmaintain electrical equipment good clean work order report defect occurliaise 

In [15]:
## Again, the description is about a Domestic role therefore, the label and the prediction are not corresponding to the meaning

### Obtain the highest probability of the predictions for each description

In [31]:
## train set 
import numpy as np
predictions_proba = svm_model.predict_proba(red_X_train).max(axis=1) 
np.sort(predictions_proba, axis=0)

array([0.27336701, 0.28961633, 0.29175376, 0.30665004, 0.31527047,
       0.3198286 , 0.32162093, 0.32551582, 0.32888284, 0.33291506,
       0.33353216, 0.33497074, 0.3357139 , 0.34233846, 0.34742793,
       0.35028844, 0.35029578, 0.35424438, 0.35842585, 0.35954686,
       0.36003666, 0.36290417, 0.36313769, 0.37498966, 0.37734999,
       0.38923607, 0.39168857, 0.39236152, 0.3958662 , 0.39943088,
       0.40486093, 0.40629149, 0.4077012 , 0.4110441 , 0.41132163,
       0.41367923, 0.4272912 , 0.42730473, 0.43050115, 0.43062008,
       0.43077273, 0.43460762, 0.43609231, 0.43820666, 0.44084998,
       0.44838586, 0.45256526, 0.45257393, 0.45599736, 0.4563747 ,
       0.45679049, 0.4599527 , 0.4599527 , 0.4599527 , 0.4599527 ,
       0.46299133, 0.46309591, 0.46581066, 0.46615363, 0.47146463,
       0.47355016, 0.47483001, 0.47519577, 0.47787459, 0.48019866,
       0.48161325, 0.48310859, 0.48518364, 0.48564162, 0.48597272,
       0.48645823, 0.48654484, 0.4868851 , 0.48700342, 0.48862

In [32]:
## test set
predictions_proba = svm_model.predict_proba(red_X_test).max(axis=1) 
np.sort(predictions_proba, axis=0)

array([0.2699256 , 0.31220316, 0.33094644, 0.33577464, 0.35166765,
       0.35994917, 0.36602158, 0.37233363, 0.37710843, 0.38367386,
       0.39004446, 0.39733909, 0.40173566, 0.41147301, 0.4167305 ,
       0.42090638, 0.43359377, 0.43533753, 0.44249372, 0.44361956,
       0.44627275, 0.4599527 , 0.47917537, 0.48725228, 0.48867977,
       0.49861346, 0.51244625, 0.51380239, 0.51411404, 0.51555304,
       0.52101913, 0.52808349, 0.53446581, 0.53770203, 0.54020486,
       0.54260531, 0.55892209, 0.55962783, 0.56072805, 0.57823266,
       0.58934742, 0.59847426, 0.60013053, 0.60437999, 0.60911238,
       0.61540901, 0.61988116, 0.62049999, 0.62283127, 0.62573241,
       0.62853658, 0.63037958, 0.63206903, 0.63501116, 0.64257214,
       0.64273088, 0.64467473, 0.65099831, 0.65131248, 0.65922554,
       0.66065531, 0.66065531, 0.6696076 , 0.67263067, 0.67321013,
       0.67814594, 0.68083493, 0.68266726, 0.68268   , 0.68524252,
       0.68751677, 0.70328675, 0.70870073, 0.70870391, 0.71208

### So, I will consider as valid outputs texts with Probability over 0.60. I will use it as a threshold for my application. Observations with a Probability lower than 0.60 will not consider as valid for predictions